torch_geometric/loader/dynamic_batch_sampler.py

```diff
@@ -0,0 +1,108 @@
+    by the sampler.
+
+    **Usage:**
+
+    .. code-block:: python
+
+        from torch_geometric.loader import DataLoader, DynamicBatchSampler
+
+        batch_sampler = DynamicBatchSampler(dataset, max_num=10000)
+        loader = DataLoader(dataset, batch_sampler=sampler, ...)
+
+    Args:
+        data_source (Dataset): dataset to sample from, need this so that
+            potential samples can be sized to create the dynamic batch.
+        max_num (int): size of mini-batch to aim for in number of nodes or
+            edges.
+        mode (str, optional): `nodes` or `edges` to measure batch size.
+            (default: `nodes`)
+        shuffle (bool, optional): set to ``True`` to have the data reshuffled
+            at every epoch (default: ``False``).
+        skip_too_big (bool, optional): set to ``True`` to skip samples which
+            can't fit in a batch by itself. (default: ``False``).
+        num_steps (int, optional): The number of mini-batches to draw for a
+            single epoch. If set to :obj:`None`, will iterate through all the
+            underlying data, but __len__ will be :obj:`None` since it will be
+            ambiguous. (default: :obj:`None`)
+    """
+    def __init__(self, data_source: Dataset, max_num: int, mode: str = 'nodes',
+                 shuffle: bool = False, skip_too_big: bool = False,
+                 num_steps: Optional[int] = None):
+        if not isinstance(max_num, int) or max_num <= 0:
+            raise ValueError("`max_num` should be a positive integer value, "
+                             "but got max_num={max_num}.")
+        if mode not in ['nodes', 'edges']:
+            raise ValueError("`mode` choice should be either "
+                             f"`nodes` or `edges, not {mode}.")
+
+        self.max_num = max_num
+        self.data_source = data_source
+        self.shuffle = shuffle
+        self.skip_too_big = skip_too_big
+        if num_steps is None:
+            num_steps = len(data_source)
+        self.num_steps = num_steps
+
+        self.mode = f"num_{mode}"
+
+    def __iter__(self) -> Iterator[List[int]]:
+        batch = []
+        batch_n = 0
+        num_steps = 0
+        current_place = 0
+
+        if self.shuffle:
+            self.idxs = torch.randperm(len(self.data_source), dtype=torch.long)
+        else:
+            self.idxs = torch.arange(len(self.data_source), dtype=torch.long)
+
+        # Main iteration loop
+        while (current_place < len(self.data_source)
+               and num_steps < self.num_steps):
+            # Fill batch
+            for idx in self.idxs[current_place:]:
+                # Size of sample
+                n = getattr(self.data_source[idx], self.mode)
+
+                if batch_n + n > self.max_num:
+                    if batch_n == 0:
+                        if self.skip_too_big:
+                            continue
+                        else:
+                            raise RuntimeError(
```

Comment:
Won‘t we want to skip instead?

Reply:
I don't quite understand this comment. Essentially the purpose of that bit is that if the batch is empty and the next sample will overfill it - either skip that sample or raise an error.
If it's not empty, currently it outputs the batch and then attempts to use that sample in the next batch.

Reply:
That makes sense! Thanks!

Reply:
I looked at it again and feel that we should output a warning here and still add that data object in (as a single example per batch). Currently, we only have the option to skip or to error out - and neither of this may be what the user desires.

Reply:
Yes that probably sounds like a better idea actually. So instead of error, just throw a warning and add the example anyway? To clarify, when this happens there would only be a single example in the mini-batch, and `data.num_nodes` will be greater than the `max_num` specified for that one mini-batch.

That sounds reasonable to me.

Reply:
Super :)